### ISSUE TITLE
Cleanup distributions: `FloatDistribution` and `IntDistribution`

### ISSUE THREAD

**Issue by himkt** [optuna/optuna#2941]
## Motivation

There are six classes that represent distributions in Optuna. For categorical parameters, `CategoricalDistribution` corresponds. For integer parameters, `IntUniformDistribution` and `IntLogUniformDistribution`. `UniformDistribution`, `LogUniformDistribution`, and `DiscreteUniformDistribution` for floating-point parameters.

`IntLoguniformDistribution` can be seen as a special case of `IntUniformDistribution`. As well, `LogUniformDistribution` and `DiscreteUniformDistribution` as of `UniformDistribution`.

In this issue, we will work on cleaning up distributions to three classes `CategoricalDistribution`, `IntDistribution`, and `FloatDistribution` by renaming and unifying the existing distributions. We think it improves the user experience because suggest APIs (`suggest_categorical`, `suggest_int`, `suggest_float`) would simply correspond to distribution names.

## Description

- [x] Introduce `FloatingDistribution` and `IntDistribution` #3063 
- [x] Add `IntDistribution` and `FloatDistribution` to the place where conditional branching is done by `UniformDistribution` etc #3111 #3126
- [x] Implement migration from current storage #3113 
- [x] Deprecate distributions mentioned in the description #2990 
- [x] Replace `UniformDistribution` etc to `IntDistribution` and `FloatDistribution` where instances of the distributions are created #3166 #3181
- [x] Update old distributions such that they inherit from `IntDistribution` and `FloatDistribution` #3244 
- [x] Automatically convert deprecated distribution if it is provided by user #3420 
- [x] Fix distribution compatibility for linear and logarithmic distribution #3444

**Comment by keisuke-umezawa** [comment]
detail considerations for `Determine and implement migration from current storage`

1. Prepare SQL migration script for [distribution_json](https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/models.py#L346). ref: [migration script](https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/alembic/versions/v2.4.0.a.py)
2. Modify check compatibility such as [_check_compatibility_with_previous_trial_param_distributions](https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/models.py#L357)

**Comment by himkt** [comment]
@nyanhi @keisuke-umezawa
Thank you so much for working on this issue.
In #2939, @xadrianzetx and I will work on a task to allow `suggest_float` to add a parameter (`step` and `log`) during the optimization. I think it needs `FloatDistribution`, which unifies floating-point distributions therefore I came here to ask if we can collaborate.

As I understand, the current status and further tasks look like following:
<img src="https://user-images.githubusercontent.com/5164000/140647509-299c6c42-303a-437a-b3e6-0a9351702d49.png" width="600">

After instantiating `FloatDistribution` instead of the current float distributions, we can add `step` or `log` during an optimization (like `IntUniformDistribution`).

@nyanhi Would you continue this issue after #3063?
I think tasks inside a rectangle can be worked on in parallel (by task or by type of distribution).
And I'd be happy to collaborate with you to move forward.
For example, I'm wondering if we can work on the tasks by the following assignment (just an example):

- @himkt or @keisuke-umezawa: create a migration script
- @nyanhi: use `FloatDistribution` in codebase
- @xadrianzetx: use `IntDistribution` in codebase

@nyanhi @xadrianzetx What do you think? I'd like to know whether you're interested in.
(and please ask if you have any questions)

**Comment by xadrianzetx** [comment]
>nyanhi: use `FloatDistribution` in codebase
xadrianzetx: use `IntDistribution` in codebase

I think this should be other way around to keep in line with #2939, right? Otherwise, looks great, I'm in!

**Comment by nyanhi** [comment]
Same here, let me work on the `IntDistribution`!

**Comment by himkt** [comment]
Thank you so much for the quick response! I'm happy to hear. 😻 

> I think this should be other way around to keep in line with #2939, right?

I just dropped the example for the assignment without the strong opinion because you both have considerable knowledge (xadrianzetx for `suggest_float` and nyanhi for `FloatDistribution` itself). But, yeah, let's work on by the suggested assign!

**Comment by xadrianzetx** [comment]
I'm scoping the work needed to add new distributions to various places in codebase. Stuff like `suggest` APIs, tests and examples where we create instances of distributions look fine, but how about places where we branch based on special case of distribution, such as
https://github.com/optuna/optuna/blob/2b0be2c1cf3ee668b3129b105a01cf172acae5b8/optuna/samplers/_tpe/parzen_estimator.py#L289-L324

In those places, where we aim to add (or switch to?) new distribution, the logic would go something like
```python
elif isinstance(distribution, distributions.FloatDistribution): 
  if distribution.log:
    # LogUniform
  if distribution.step is not None:
    # DiscreteUniform
  else:
    # Uniform
```
I'm asking since I'm not sure about using attributes to determine the case here (readability reasons). Maybe having some additional methods available to identify distributions would be a good idea? E.g.

```python
class FloatDistribution()
  [...]
  def is_discrete(self):
    return self.step is not None

  def is_log(self):
    return self.log
``` 
Let me know if this is overkill or not.

**Comment by himkt** [comment]
I think your first thought is just enough for this time. After introducing new distributions were completed, I want to modify codebase more elegant and readable with you.

```python
elif isinstance(distribution, distributions.FloatDistribution): 
  if distribution.log:
    # LogUniform
  if distribution.step is not None:
    # DiscreteUniform
  else:
    # Uniform
```

@nyanhi, what do you think about it? It's nice to have a discussion together because you will work on the similar task.

**Comment by himkt** [comment]
@xadrianzetx I think your first thought is just enough for this time. After introducing new distributions were completed, I want to modify codebase more elegant and readable with you.

```python
elif isinstance(distribution, distributions.FloatDistribution): 
  if distribution.log:
    # LogUniform
  if distribution.step is not None:
    # DiscreteUniform
  else:
    # Uniform
```

@nyanhi, what do you think about it? It's nice to have a discussion together because you will work on the similar task.

**Comment by nyanhi** [comment]
I agree that the first choice seems reasonable for now. 

My naive idea is to add a method `BaseDistribution.transform` works like below so that distribution classes can internally make necessary change to the given np.ndarray `samples`, depending on its attributes `log` and `step`. 
However, please kindly note that I am saying this without fully considering whether this is in line with the design of the distribution class at this moment (sorry if it's too random thought...).

```
transformed[param_name] = distribution.transform(samples)
```

**Comment by HideakiImamura** [comment]
All tasks in this item have been finished. Let me close this PR. Thank you for your all of contributions! @xadrianzetx @nyanhi @himkt
